Runtime/Triangulator.cs

```diff
@@ -173,187 +234,217 @@ public class InputData
+                localPositions = output.Positions.IsCreated ? output.Positions : new NativeList<float2>(input.Positions.Length, Allocator.Temp);
+                localPositions.ResizeUninitialized(input.Positions.Length);
+                if (preprocessor == Preprocessor.PCA || preprocessor == Preprocessor.COM) {
+                    localTransformation = preprocessor == Preprocessor.PCA ? CalculatePCATransformation(input.Positions) : CalculateLocalTransformation(input.Positions);
+                    localTransformation.Transform(input.Positions, localPositions.AsArray());
+                    if (input.HoleSeeds.IsCreated) {
+                        localHoles = new NativeArray<float2>(input.HoleSeeds.Length, Allocator.Temp);
+                        localTransformation.Transform(input.HoleSeeds, localHoles);
+                    } else {
+                        localHoles = default;
+                    }
+                } else if (preprocessor == Preprocessor.None) {
+                    localPositions.CopyFrom(input.Positions);
+                    localHoles = input.HoleSeeds;
+                    localTransformation = AffineTransform2D.identity;
+                } else {
+                    throw new System.ArgumentException();
+                }
+            }
+
+            public void Execute() {
+                output.Status.Value = Status.OK;
+                output.Triangles.Clear();
+                output.Positions.Clear();
+
+                MarkerPreProcess.Begin();
+                PreProcessInput(preprocessor, input, output, out var localPositions, out var localHoles, out var localTransformation);
+                MarkerPreProcess.End();
+
+                if (validateInput) {
+                    MarkerValidateInput.Begin();
+                    ValidateInput(localPositions.AsArray(), input.ConstraintEdges);
+                    MarkerValidateInput.End();
+                }
+
+                MarkerDelaunayTriangulation.Begin();
+                var halfEdges = new NativeList<int>(localPositions.Length, Allocator.Temp);
+                var triangles = output.Triangles;
+                var circles = new NativeList<Circle>(localPositions.Length, Allocator.Temp);
+                new DelaunayTriangulationJob {
+                    status = output.Status,
+                    positions = localPositions.AsArray(),
+                    triangles = triangles,
+                    halfedges = halfEdges,
+                    hullStart = int.MaxValue,
+                    c = float.MaxValue,
+                }.Execute();
+                MarkerDelaunayTriangulation.End();
+
+                NativeList<Edge> internalConstraints = default;
+                if (input.ConstraintEdges.IsCreated) {
+                    internalConstraints = new NativeList<Edge>(localPositions.Length, Allocator.Temp);
```

Comment:
I encountered several errors while running tests locally on my machine.

![image](https://github.com/andywiecko/BurstTriangulator/assets/15237586/21f374f9-00db-461b-8eec-bb2938777ad4)

Most of the errors is related to `Burst` foreach bug

```csharp
InvalidOperationException: The Unity.Collections.NativeList`1[System.Int32] has been declared as [WriteOnly] in the job, but you are reading from it.
Unity.Collections.LowLevel.Unsafe.AtomicSafetyHandle.CheckReadAndThrowNoEarlyOut (Unity.Collections.LowLevel.Unsafe.AtomicSafetyHandle handle) (at <099b34bda0694abe92051a650b8f9fc9>:0)
Unity.Collections.LowLevel.Unsafe.AtomicSafetyHandle.CheckReadAndThrow (Unity.Collections.LowLevel.Unsafe.AtomicSafetyHandle handle) (at <099b34bda0694abe92051a650b8f9fc9>:0)
Unity.Collections.NativeArray`1[T].CheckElementReadAccess (System.Int32 index) (at <099b34bda0694abe92051a650b8f9fc9>:0)
Unity.Collections.NativeArray`1[T].get_Item (System.Int32 index) (at <099b34bda0694abe92051a650b8f9fc9>:0)
andywiecko.BurstTriangulator.Triangulator+ConstrainEdgesJob.CollectIntersections (andywiecko.BurstTriangulator.Triangulator+Edge edge) (at C:/Users/Andrzej/Documents/andywiecko/BurstTriangulator/Runtime/Triangulator.cs:1299)
andywiecko.BurstTriangulator.Triangulator+ConstrainEdgesJob.TryApplyConstraint (andywiecko.BurstTriangulator.Triangulator+Edge c) (at C:/Users/Andrzej/Documents/andywiecko/BurstTriangulator/Runtime/Triangulator.cs:1131)
andywiecko.BurstTriangulator.Triangulator+ConstrainEdgesJob.Execute () (at C:/Users/Andrzej/Documents/andywiecko/BurstTriangulator/Runtime/Triangulator.cs:1110)
andywiecko.BurstTriangulator.Triangulator+TriangulationJob.Execute () (at C:/Users/Andrzej/Documents/andywiecko/BurstTriangulator/Runtime/Triangulator.cs:382)
```

To address this issue, the simplest solution is to replace `NativeList<Edge>` with `NativeArray<Edge>` and call `AsReadOnly()` within the loop

```csharp
                var internalConstraints = default(NativeArray<Edge>);
                ...
                foreach (var c in internalConstraints.AsReadOnly())
                {
                    TryApplyConstraint(c);
                }
```

PS
Additionally, after my upcoming PR with `constraintHalfedges` generation, which I mentioned before, this `internalConstraints` array can be removed entirely from the codebase, thanks to your job unification.

Reply:
I changed it to an array. However, I had to make some changes to the `RefineMeshJob`, as it was still required it to be a list. But I think the new version is better and faster anyway.
I'm not quite sure why certain if-statements had to be the way they are, but it matches the behavior of your previous code.